src/representations/representation_manager_spherical_invariants.hh

```diff
@@ -683,14 +784,328 @@ namespace rascal {
+                   ++cartesian_idx) {
+                size_t cartesian_offset_n{cartesian_idx * this->max_radial};
+                norm_grad_neigh[cartesian_idx] +=
+                    (coef_grad_neigh
+                         .block(cartesian_offset_n, 0, this->max_radial, 1)
+                         .array() *
+                     coef_by_type.array())
+                        .sum();
+              }
+            }
+            norm_grad_neigh *= coefficients_norm_inv3;
+
+            for (const auto & el : coefficients) {
+              element_type[0] = el.first[0];
+              auto && coef{el.second};
+              auto && soap_neigh_gradient_by_type{
+                  soap_neigh_gradient[element_type]};
+              for (size_t cartesian_idx{0}; cartesian_idx < 3;
+                   ++cartesian_idx) {
+                size_t cartesian_offset_n{cartesian_idx * this->max_radial};
+                soap_neigh_gradient_by_type.block(cartesian_offset_n, 0,
+                                                  this->max_radial, 1) -=
+                    coef * norm_grad_neigh[cartesian_idx];
+              }
+            }
+          }  // for (auto neigh : center)
+        }    // if (this->normalize)
+      }      // if (this->compute_gradients)
+    }        // for (auto center : this->structure_manager)
+  }
+
+  template <class Mngr>
+  void RepresentationManagerSphericalInvariants<Mngr>::compute_bispectrum() {
+    using internal::enumValue;
+    using internal::SphericalInvariantsType;
+
+    auto precomputation{downcast_spherical_invariants_precompute<
+        SphericalInvariantsType::BiSpectrum>(
+        this->precompute_spherical_invariants[enumValue(
+            SphericalInvariantsType::BiSpectrum)])};
+    auto & wigner_w3js{precomputation->wigner_w3js};
+
+    rep_expansion.compute();
+
+    this->initialize_percenter_bispectrum_soap_vectors();
+
+    auto & expansions_coefficients{rep_expansion.expansions_coefficients};
+
+    using complex = std::complex<double>;
+
+    // factor that takes into acount the missing equivalent off diagonal
+    // element with respect to the key (or species) index
+    double mult{1.0};
+    Key_t trip_type{0, 0, 0};
+    internal::SortedKey<Key_t> triplet_type{trip_type};
+    for (auto center : this->structure_manager) {
+      auto & coefficients{expansions_coefficients[center]};
+      auto & soap_vector{this->soap_vectors[center]};
+
+      for (const auto & el1 : coefficients) {
+        triplet_type[0] = el1.first[0];
+        auto & coef1{el1.second};
+        for (const auto & el2 : coefficients) {
+          triplet_type[1] = el2.first[0];
+          auto & coef2{el2.second};
+          for (const auto & el3 : coefficients) {
+            triplet_type[2] = el3.first[0];
+            auto & coef3{el3.second};
+            // triplet multiplicity
+            // all the same
+            if (triplet_type[0] == triplet_type[1] &&
```

Comment:
I am not sure what happens here, but it seems like a hot loop and there are a lot of "if's" inside. Could you clarify? Is that just normalization? Could this loop benefit from sorting?

Reply:
there is very little that can be done here without drastic changes that would hinder readability. Let's keep the improvements to latter

Reply:
The maths is complicated enough as it is, so it looks like we're going for something that works for now and can be optimized later.